docs/references/security/rust-canister-development-security-best-practices.md

```diff
@@ -183,43 +183,157 @@ A canister could be rendered unusable so it could never be upgraded again e.g. d
 
 ## Inter-Canister Calls and Rollbacks
 
-### Don’t panic after await and don’t lock shared resources across await boundaries
+### Message Execution Basics
+
+To understand the issues around async inter-canister calls, one needs to understand a few properties about message execution. This is also explained in the [community conversation on Security Best Practices](https://www.youtube.com/watch?v=PneRzDmf_Xw&list=PLuhDt1vhGcrez-f3I0_hvbwGZHZzkZ7Ng&index=2&t=4s). A _message_ is a set of consecutive instructions that a subnet executes for a canister. The following properties are essential: 
+
+**Property 1**: Only a single message is processed at a time. So message execution is sequential, and never parallel.
+
+**Property 2**: Each call (query / update) triggers a message. When an inter-canister call is made, the code after the call (the callback) is executed as a separate message. 
+
+For example, consider the following code in Motoko: 
+
+
+![example_highlighted_code](_attachments/example_highlighted_code.png)
+
+The first message that is executed here are lines 2-3, until the inter-canister call is made using the `await` syntax (orange box). The second message are lines 3-5: when the inter-canister call returns (blue box). We call this part the _callback_ of the inter-canister call. The two messages involved in this example will always be scheduled sequentially. 
+
+**Property 3**: Messages from interleaving calls have no reliable execution ordering.
+
+Let's consider the above example code again, and assume the method `example` is called twice in parallel, the resulting calls being Call 1 and Call 2. The following illustration shows two possible message orderings. On the left, the first call's messages are scheduled first, and only then the second call's messages are executed. On the right, we see another possible message scheduling, where the first message of each call are executed first. 
+
+![example_orderings](_attachments/example_orderings.png)
+
+**Property 4**: On a trap / panic, modifications to the canister state for the current message are not applied.
+
+For example, if a trap in the second message (blue box) of the above example occurs, any canister state changes for that message, even earlier in the blue box, are not applied. 
```

Comment:
```suggestion
For example, if a trap in the second message (blue box) of the above example occurs, any canister state changes as a result of that message, even earlier in the blue box, are discarded.
```

Reply:
I'd also add a note that changes of the earlier messages in the same call context *are* persisted. So we don't have method-level rollback, which means that intermediate states will get persisted unless there's cleanup code somewhere that tries to fix things.

Reply:
OK, I see that you discuss it in detail in the next subsection, I'd still emphasize that the state changes of earlier callbacks (orange in your example) are not rolled back